[fx] add type annotations

zdevito opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#43083 [fx] add type annotations**
* #43082 [fx] enabling typechecking of fx files

This adds type annotations to all classes, arguments, and returns
for fx. This should make it easier to understand the code, and
encourage users of the library to also write typed code.

Differential Revision: [D23145853](https://our.internmc.facebook.com/intern/diff/D23145853)

zdevito committed:
[fx] add type annotations

This adds type annotations to all classes, arguments, and returns
for fx. This should make it easier to understand the code, and
encourage users of the library to also write typed code.

[ghstack-poisoned]

zdevito committed:
Update on "[fx] add type annotations"


This adds type annotations to all classes, arguments, and returns
for fx. This should make it easier to understand the code, and
encourage users of the library to also write typed code.

Differential Revision: [D23145853](https://our.internmc.facebook.com/intern/diff/D23145853)

[ghstack-poisoned]

ezyang commented on torch/fx/graph.py:
```diff
@@ -61,16 +62,19 @@ def map_arg(a, fn):
 
 class Graph:
```
NB: you can declare

```
nodes: List[Node]
_used_names: Dict[str, int]
```

up outside of `__init__` which I think looks nice and is nicely documentary

ezyang reviewed (approved):
Good stuff.

ezyang commented:
I don't see changes to the mypy config to ensure we are strictly testing this in CI, would be a good thing to do eventually!

zdevito commented:
@ezyang re: mypy config, I was under the impression deleting `# type: ignore` from the top of the files enabled it. What do I have to do to turn it on in CI?

jamesr66a commented:
Yeah `# type: ignore` disabled some failing mypy checks in CI, so I'm assuming these are getting checked? Unless there's some other CI configs that should be kicking in as well

zdevito committed:
Update on "[fx] add type annotations"


This adds type annotations to all classes, arguments, and returns
for fx. This should make it easier to understand the code, and
encourage users of the library to also write typed code.

Differential Revision: [D23145853](https://our.internmc.facebook.com/intern/diff/D23145853)

[ghstack-poisoned]

ezyang commented:
> I was under the impression deleting # type: ignore from the top of the files enabled it. What do I have to do to turn it on in CI?

You are getting type checking, but it's the "lax" typechecking that is configured by the default `mypy.ini`. This, among other things, won't error if people fail to add type signatures to functions.

Because FX imports torch, it's not going to be so simple to enable strict type checking; you'll need to list all of the FX files, and then also configure mypy with `--follow-imports=silent` so it doesn't bleat about lack of full typing in torch. This is why I think you should do it at some point, but you don't have to block this PR on it.

zdevito committed:
Update on "[fx] add type annotations"


This adds type annotations to all classes, arguments, and returns
for fx. This should make it easier to understand the code, and
encourage users of the library to also write typed code.

Differential Revision: [D23145853](https://our.internmc.facebook.com/intern/diff/D23145853)

[ghstack-poisoned]
